File: rag-langchain-python/contextual_query_retriever.py
from typing import List
from langchain_core.documents import Document
from langchain_core.retrievers import (
    BaseRetriever,
)
from marklogic import Client


class ContextualQueryRetriever(BaseRetriever):
    client: Client

    @classmethod
    def create(cls, client: Client):
        return cls(client=client)

    def _get_relevant_documents(
        self,
        chat_context: object,
    ) -> List[Document]:
        words = []
        for word in chat_context["question"].split():
            if len(word) > 2:
                words.append(word.lower().replace("?", ""))
        term_query = {"term-query": {"text": words}}

        chat_context["contextual_query"]["query"]["queries"].append(term_query)

        print(f"Searching with query: {chat_context['contextual_query']}")
        results = self.client.documents.search(
            query=chat_context["contextual_query"],
            page_length=10,
            collections=["events"],
        )

        print(f"Count of MarkLogic documents sent to the LLM: {len(results)}")
        for result in results:
            print(f"URI: {result.uri}")
        return map(lambda doc: Document(page_content=doc.content["transcript"]), results)

File: rag-langchain-js/contextualQueryRetriever.js
import { BaseRetriever } from "@langchain/core/retrievers";
import { Document } from "@langchain/core/documents";

// This example is based on the LangChain guide at https://js.langchain.com/v0.1/docs/modules/data_connection/retrievers/custom/
// and the example at https://github.com/langchain-ai/langchainjs/blob/main/examples/src/retrievers/custom.ts
export class ContextualQueryRetriever extends BaseRetriever {
  lc_namespace = ["langchain", "retrievers"];
  marklogicClient;

  constructor(fields) {
    super(fields);
    this.marklogicClient = fields.marklogicClient;
  }

  #buildWordsArray(chatQuestion) {
    const words = [];
    chatQuestion.split(" ").forEach(word => {
      if (word.length > 2) words.push(word.toLowerCase().replace("?", ""))
    })
    return words;
  }

  async _getRelevantDocuments(
    questionContext, _callbacks,
  ) {
    const words = this.#buildWordsArray(questionContext.question);
    const wordQuery = {"term-query": {"text": words}};
    const contextualQuery = questionContext.contextualQuery;
    contextualQuery["query"]["queries"].push(wordQuery)
    console.log(JSON.stringify(contextualQuery))

    const results = await this.marklogicClient.documents.query(contextualQuery).result((queryResults) => {
      return queryResults;
    });

    const documents = [];
    results.forEach((result) => {
      console.log(result["uri"])
      documents.push(
        new Document({
          pageContent: result['content']['transcript'],
          metadata: {"uri": result["uri"]}
        })
      )
    });
    return documents;
  }
}
